app/custom_auth.py

```diff
@@ -0,0 +1,74 @@
+from django.contrib import auth
+from django.conf import settings
+from django.contrib.auth.models import User
+
+class CustomSSOMiddleware:
+    """
+    A custom middleware that allows SSO via HTTP header passed in the request
+    The name of HTTP header is defined in the settings.py by the variable HEADER_CUSTOM_USER_NAME  
+
+    
+    (1) If HTTP header with the user name is not passed no further processing is done by the middleware.
+    (2) If HTTP header with the user name is passed then:
+        (2.1) Authenciate the user if already exist in the database. 
+        (2.2) Create user if does not exist in the database and then autenicate the user. 
+        (2.3) User is now signed in and session cookies are set.  
+    ...
+    """
+    def __init__(self, get_response):
+        print ('LOG: Inside the custom middleware!   __init__')
+        self.get_response = get_response
+     
+    def __call__(self, request):
+        print ('LOG: Inside the custom middleware!   __call__')
+        
+        header =  settings.HEADER_CUSTOM_USER_NAME
+        username= ""
+
+        try:
+            username = request.META[header]
+        except:
+         
+            print ("LOG: User name header not found: " + header)
+            response = self.get_response(request)
+            return response
+
+        print( 'LOG: Username provided by the header :' + username)
+        is_user_exists = username_present(username)
+        user = None
+
+        if is_user_exists == True:
+            user = auth.authenticate(remote_user=username)
+        else:
+            print ('LOG: User is not defined: '+username)
+            if settings.CREATE_USER == 'TRUE':
+                user = create_user(username)
+                if user is not None:
+                   user = auth.authenticate(remote_user=username)
+
+        if user:
+            request.user = user
+            auth.login(request, user)
+
+        response = self.get_response(request)
+
+        return response
+
+def username_present(username):
+    if User.objects.filter(username=username).exists():
+        return True
+
+def create_user(username):
+
+    print ('LOG: create_user called: ' +username)
+    user = None
+    try:
+        email = username + "@noreply.com"
+        user = User.objects.create_user(username=username,email=email)
+        user.is_superuser=settings.IS_SUPER
+        user.save()
+
+        return user
+    except:
```

Comment:
Given that we don't have any good action we can take to recover from the exception, might as well bubble up the exception or handle it at the top level in the middleware. Returning `None` and checking for that instead of using exceptions is somewhat discouraged in the Python community. [See EAFP vs LBYL](https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/).